src/Orleans/Messaging/SqlMembershipTable.cs

```diff
@@ -21,489 +21,229 @@ MIT License
-
-                        return result > 0;
-                    }
-                }
-                return false;
+                return database.AllMembershipDataAsync(deploymentId);                
             }
-        }
-
-        async Task IMembershipTable.UpdateIAmAlive(MembershipEntry entry)
-        {
-            // Update, but only the IAmAlive column.
-            using (var conn = new SqlConnection(connectionString))
+            catch(Exception ex)
             {
-                conn.Open();
-
-                var command = new SqlCommand(UPDATE_ALIVE_TIME);
-                UpdateIAmAliveTime(entry, command, this.deploymentId);
-                command.Connection = conn;
-
-                await command.ExecuteNonQueryAsync();
+                logger.Verbose("SqlMembershipTable.ReadAll failed: {0}", ex);
+                throw;
             }
         }
 
-        public async Task DeleteMembershipTableEntries(string deployId)
-        {
-            using (var conn = new SqlConnection(connectionString))
-            {
-                conn.Open();
-
-                using (var tx = conn.BeginTransaction())
-                {
-                    var command = new SqlCommand(DELETE_TABLE_ENTRIES);
-                    command.Parameters.Add(new SqlParameter { ParameterName = "@deploymentid", DbType = DbType.String, Value = deployId });
-                    command.Connection = conn;
-                    command.Transaction = tx;
 
-                    await command.ExecuteNonQueryAsync();
-
-                    command = new SqlCommand(DELETE_VERSION_ROW);
-                    command.Parameters.Add(new SqlParameter { ParameterName = "@deploymentid", DbType = DbType.String, Value = deployId });
-                    command.Connection = conn;
-                    command.Transaction = tx;
-
-                    await command.ExecuteNonQueryAsync();
-
-                    tx.Commit();
-                }
-            }
-        }
-        
-        internal static string NewEtag()
+        Task<bool> IMembershipTable.InsertRow(MembershipEntry entry, TableVersion tableVersion)
         {
-            return Guid.NewGuid().ToString();
-        }
+            logger.Verbose3(string.Format("SqlMembershipTable.InsertRow called with entry {0} and tableVersion {1}.", entry, tableVersion));
 
-        private async Task InitTable()
-        {
-            using (var conn = new SqlConnection(connectionString))
+            //The "tableVersion" parameter should always exist when inserting a row as Init should
+            //have been called and membership version created and read. This is an optimization to
+            //not to go through all the way to database to fail a conditional check on etag (which does
+            //exist for the sake of robustness) as mandated by Orleans membership protocol.
+            //Likewise, no update can be done without membership entry.
+            if(tableVersion == null || entry == null)
```

Comment:
These are sanity checks. No need to go over the wire all the way to the database to get a failure if there's a fatal problem. This shouldn't happen in normal, non-development scenarios, I suppose.